compiler/rustc_trait_selection/src/error_reporting/infer/mod.rs

```diff
@@ -1141,15 +1141,14 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
     #[instrument(level = "debug", skip(self, diag, secondary_span, prefer_label))]
     pub fn note_type_err(
         &self,
+        span: Span,
         diag: &mut Diag<'_>,
         cause: &ObligationCause<'tcx>,
```

Comment:
Why are we passing a *different* span than the span passed in via the obligation cause? 

This *at least* needs an explanation; especially when an argument seems to be redundant in 19 out of 20 calls, and *only* differs, but there's literally no documentation to explain *why* it's like that. I find that adding new arguments like this with basically no rhyme or reason to these very pervasive functions (note_type_err is the entrypoint to basically all type error reporting!) to really harm maintainability of these APIs :(

Reply:
An idea is to rename this to `override_span: Option<Span>`, and move it to the *end* of the arg list (or perhaps just behind `seconary_span`, and add a doc comment explaining that passing `None` will use the obligation cause's span. Then most of the callsites will just need to pass `None` rather than redundantly passing both the obligation cause and the obligation cause's span as two args.

Reply:
Let me know what you think.